src/provider/non_hiding_kzg.rs

```diff
@@ -334,6 +376,61 @@ where
     let pairing_result = E::multi_miller_loop(pairing_input_refs.as_slice()).final_exponentiation();
     Ok(pairing_result.is_identity().into())
   }
+
+  /// Verifies that `value_i` is the evaluation at `x_i` of the polynomial
+  /// `poly_i` committed inside `comm`.
+  // This is a naive approach
+  // TODO: to implement the more efficient batch verification algorithm
+  #[allow(dead_code, clippy::unnecessary_wraps)]
+  pub fn batch_verify<R: RngCore + CryptoRng>(
+    verifier_params: impl Borrow<KZGVerifierKey<E>>,
+    multi_commitment: &[UVKZGCommitment<E>],
+    points: &[E::Fr],
+    values: &[UVKZGEvaluation<E>],
+    batch_proof: &[UVKZGProof<E>],
+    rng: &mut R,
+  ) -> Result<bool, NovaError> {
+    let verifier_params = verifier_params.borrow();
+
+    let mut total_c = <E::G1>::identity();
+    let mut total_w = <E::G1>::identity();
+
+    let mut randomizer = E::Fr::ONE;
+    // Instead of multiplying g and gamma_g in each turn, we simply accumulate
+    // their coefficients and perform a final multiplication at the end.
+    let mut g_multiplier = E::Fr::ZERO;
+    zip_with_for_each!(
+      into_iter,
+      (multi_commitment, points, values, batch_proof),
+      |c, z, v, proof| {
+        let w = proof.proof;
+        let mut temp = w.mul(*z);
+        temp += &c.0;
+        let c = temp;
+        g_multiplier += &(randomizer * v.0);
+        total_c += &c.mul(randomizer);
+        total_w += &w.mul(randomizer);
+        // We don't need to sample randomizers from the full field,
+        // only from 128-bit strings.
+        randomizer = E::Fr::from_u128(rand::Rng::gen::<u128>(rng));
+      }
+    );
+    total_c -= &verifier_params.g.mul(g_multiplier);
+
+    let mut affine_points = vec![E::G1Affine::identity(); 2];
+    E::G1::batch_normalize(&[-total_w, total_c], &mut affine_points);
+    let (total_w, total_c) = (affine_points[0], affine_points[1]);
+
+    let result = E::multi_miller_loop(&[
```

Comment:
As Adrian mentioned (and I then could also convince myself by observing the implementation), this form of pairings check is more optimal, comparing to:

```
let left = E::pairing(..., ...);
let right = E::pairing(..., ...);
assert_eq!(left, right);
```

as it saves one `multi_miller_loop` and one `final_exponentiation` call. However that requires some manipulating with input parameters, which differs from what is necessary to pass to `pairing` couple. For example, given vanilla KZG10 specification, described [here](https://risencrypto.github.io/Kate/), the final check is specified like:

```
e(C_q, a⋅G2−b⋅G2) == e(C_f−G1⋅c, G2) 
```

which has following Rust equivalent:
```
let g1 = srs.powers_of_g[0];
let g2 = srs.powers_of_h[0];

let left = pairing(
  &Cq.to_affine(),
  &(srs.powers_of_h[1] - (g2 * b)).to_affine(),
);
let right = pairing(&(Cf.to_affine() - (g1 * c).to_affine()).to_affine(), &g2);
assert_eq!(left, right);
```

Alternatively (more optimally), one can write this check as:
```
// Alternatively (more optimal as it doesn't require executing final_exponentiation twice)
let term1 = (
  &(srs.powers_of_g[0] * c - Cq * b - Cf).to_affine(),
  &G2Prepared::from_affine(srs.powers_of_h[0]),
);
let term2 = (
  &Cq.to_affine(),
  &G2Prepared::from_affine(srs.powers_of_h[1]),
);
let result = multi_miller_loop(&[term1, term2]).final_exponentiation();
assert_eq!(result.is_identity().unwrap_u8(), 0x01);
```
which I derived empirically. 

I wonder if there is a systematic way that allows turning the `e() == e()` formula into more optimal one:
```
term1 = ...
term2 = ...
multi_miller_loop(term1, term2).final_exponentiation().is_identity()
```

Thanks in advance, for answering, @huitseeker !